### PR TITLE
Menu

### PR DESCRIPTION
**Pull request by R00BEL**

### CODE REVIEW

**Comment by pkarpovich** [comment]
move photos to backend 
https://docs.nestjs.com/recipes/serve-static#serve-static

File: src/components/MenuLayout.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        </Flex>
+      </Flex>
+    </>
+  );
+};
+
+const s = {
+  Container: styled(Flex)`
+    overflow: hidden;
+    width: 100%;
```
Comment (pkarpovich) [inline src/components/MenuLayout.js]: use properties from Flex component

File: src/components/MenuLayout.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import { Divider } from "./divider";
+
+const MenuLayout = ({ children }) => {
+  return (
+    <>
+      <DarkModeSwitcher />
+      <Flex direction="column" width={1} height={1}>
+        <s.Container>{children}</s.Container>
+        <Divider />
+        <Flex height={1} flex={1} width={1} justifyContent="space-around">
```
Comment (pkarpovich) [inline src/components/MenuLayout.js]: try to reuse the simple layout component. Just add property 'needMenu'

File: src/components/divider.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { styled } from "@linaria/react";
+
+import { Flex } from "./flex";
+
+export const Divider = styled(Flex)`
+  border-bottom: 1px solid var(--shadow);
+  width: 300%;
```
Comment (pkarpovich) [inline src/components/divider.js]: 300?

File: src/components/h1.js [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import { styled } from "@linaria/react";
 
 export default styled.h1`
+  margin-top: ${(props) => props.marginTop ?? "0.67em"};
```
Comment (pkarpovich) [inline src/components/h1.js]: 0.7

File: src/components/h2.js [hunk 5]
```diff
@@ -3,4 +3,14 @@ import { styled } from "@linaria/react";
 export default styled.h2`
   text-align: center;
   color: var(--main-text-color);
+
+  padding: ${(props) => props.padding ?? props.p ?? "0"};
```
Comment (pkarpovich) [inline src/components/h2.js]: Do you really need all these properties in that component?

File: src/components/img.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
+import { styled } from "@linaria/react";
+
+export const Img = styled.img`
+  width: ${(props) => (props.width ? `${100 * props.width}%` : "auto")};
```
Comment (pkarpovich) [inline src/components/img.js:2]: same as above

File: src/components/label.js [hunk 7]
```diff
@@ -2,4 +2,14 @@ import { styled } from "@linaria/react";
 
 export const Label = styled.label`
   color: var(--main-text-color);
+
+  padding: ${(props) => props.padding ?? props.p ?? "0"};
```
Comment (pkarpovich) [inline src/components/label.js]: same as above

File: src/components/menu-tabs/icon_basket.svg [hunk 8]
```diff
@@ -0,0 +1,4 @@
+<svg width="24" height="25" viewBox="0 0 24 25" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (pkarpovich) [inline src/components/menu-tabs/icon_basket.svg]: move it into assets icon folder

File: src/components/menu-tabs/icon_basket_dedicated.svg [hunk 9]
```diff
@@ -0,0 +1,4 @@
+<svg width="24" height="25" viewBox="0 0 24 25" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (pkarpovich) [inline src/components/menu-tabs/icon_basket_dedicated.svg]: same as above

File: src/components/menu-tabs/icon_menu.svg [hunk 10]
```diff
@@ -0,0 +1,3 @@
+<svg width="24" height="25" viewBox="0 0 24 25" xmlns="http://www.w3.org/2000/svg">
```
Comment (pkarpovich) [inline src/components/menu-tabs/icon_menu.svg]: same as above and check icons name '_'

File: src/components/menu-tabs/icon_menu_dedicated.svg [hunk 11]
```diff
@@ -0,0 +1,5 @@
+<svg width="24" height="25" viewBox="0 0 24 25" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (pkarpovich) [inline src/components/menu-tabs/icon_menu_dedicated.svg]: same as above

File: src/atoms/person/index.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
+import { atom } from "recoil";
+import { recoilPersist } from "recoil-persist";
+
+const { persistAtom } = recoilPersist();
+
+const personState = atom({
+  key: "personState",
+  default: 1,
+  dangerouslyAllowMutability: true,
```
Comment (pkarpovich) [inline src/atoms/person/index.js]: remove it

File: src/components/MainLayout.js [hunk 13]
```diff
@@ -18,10 +18,12 @@ const MainLayout = ({ children }) => {
 const s = {
   PageContainer: styled.div`
     height: 70vh;
+
     display: flex;
-    justify-content: center;
     align-items: center;
     flex-direction: column;
+
+    overflow-y: scroll;
```
Comment (pkarpovich) [inline src/components/MainLayout.js]: auto

File: src/components/menu-tabs/index.js [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+    expandedMenuParams,
+    menuParams,
+    menu,
+    productParams,
+  ]);
+
+  return (
+    <>
+      {menu || expandedMenu || product ? (
+        <Img>
```
Comment (pkarpovich) [inline src/components/menu-tabs/index.js]: Box

**Review by pkarpovich — CHANGES_REQUESTED** [review]

File: src/components/MenuLayout.js (hunk 1 above)
Comment (R00BEL) [reply src/components/MenuLayout.js]: fixed

File: src/components/h1.js (hunk 4 above)
Comment (R00BEL) [reply src/components/h1.js]: fixed

File: src/components/h2.js (hunk 5 above)
Comment (R00BEL) [reply src/components/h2.js]: fixed

File: src/components/label.js (hunk 7 above)
Comment (R00BEL) [reply src/components/label.js]: fixed

File: package.json [hunk 15]
```diff
@@ -25,7 +25,11 @@
     "react-konva": "^17.0.1-3",
     "react-localization": "^1.0.16",
     "react-scripts": "4.0.3",
+    "react-swipeable": "^6.1.2",
+    "recoil": "^0.3.1",
+    "recoil-persist": "^2.5.0",
     "use-image": "^1.0.7",
+    "web-vitals": "^2.0.1",
```
Comment (pkarpovich) [inline package.json]: remove it

File: src/atoms/personCount/index.js [hunk 16]
```diff
@@ -0,0 +1,12 @@
+import { atom } from "recoil";
+import { recoilPersist } from "recoil-persist";
+
+const { persistAtom } = recoilPersist();
+
+const personCountState = atom({
```
Comment (pkarpovich) [inline src/atoms/personCount/index.js]: move into basket atom

File: src/atoms/basket/index.js [hunk 17]
```diff
@@ -0,0 +1,12 @@
+import { atom } from "recoil";
+import { recoilPersist } from "recoil-persist";
+
+const { persistAtom } = recoilPersist();
+
+const productsInBasketState = atom({
+  key: "productsInBasketState",
+  default: [],
```
Comment (pkarpovich) [inline src/atoms/basket/index.js]: change to obj

File: src/components/MainLayout.js [hunk 18]
```diff
@@ -18,10 +18,12 @@ const MainLayout = ({ children }) => {
 const s = {
   PageContainer: styled.div`
     height: 70vh;
+
     display: flex;
-    justify-content: center;
```
Comment (pkarpovich) [inline src/components/MainLayout.js:22]: return this line

File: src/components/dark-mode-switcher/index.js [hunk 19]
```diff
@@ -3,8 +3,8 @@ import { styled } from "@linaria/react";
 
 import { Keys } from "utils/localStorage";
 import { useDarkMode } from "contexts/dark-mode-context";
-import DarkModeIcon from "./dark-mode-symbol.svg";
-import LightModeIcon from "./light-mode-symbol.svg";
+import DarkModeIcon from "../../assets/icons/dark-mode-switcher/dark-mode-symbol.svg";
```
Comment (pkarpovich) [inline src/components/dark-mode-switcher/index.js]: ../../../../

File: src/pages/menu/basket/components/swipe-delete/index.js [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+const SwipeDelete = ({ children, itemId, setCandidateForDeletion, index }) => {
+  const handlers = useSwipeable({
+    onSwipedLeft: () => {
+      setCandidateForDeletion({
+        indexInBasketAtom: index,
+        id: itemId,
+      });
+    },
+    onSwipedRight: () => {
+      setCandidateForDeletion((candidateForDeletion) => {
```
Comment (pkarpovich) [inline src/pages/menu/basket/components/swipe-delete/index.js]: onPreRemove

File: src/pages/menu/basket/index.js [hunk 21]
```diff
@@ -0,0 +1,381 @@
...
+
+const Basket = () => {
+  const [productsInBasketAtoms, setProductsInBasketAtoms] = useRecoilState(
+    productsInBasketState
+  );
+  const [personCountAtoms, setPersonCountAtoms] = useRecoilState(
+    personCountState
+  );
+
+  const [candidateForDeletion, setCandidateForDeletion] = useState({
```
Comment (pkarpovich) [inline src/pages/menu/basket/index.js]: preRemoveItemId

File: src/pages/menu/basket/index.js [hunk 22]
```diff
@@ -0,0 +1,381 @@
...
+  const [disable, setDisable] = useState(false);
+
+  const [, setLocation] = useLocation();
+  const [, { restaurantId, tableId }] = useRoute(Routes.BASKET);
+
+  const {
+    strings: { basket: translations },
+  } = useTranslation();
+
+  const sum = useMemo(() => {
```
Comment (pkarpovich) [inline src/pages/menu/basket/index.js]: refactor

File: src/pages/menu/basket/index.js [hunk 23]
```diff
@@ -0,0 +1,381 @@
...
+      productsInBasketAtoms[0].restaurantId !== restaurantId
+    ) {
+      setProductsInBasketAtoms([]);
+    }
+
+    // eslint-disable-next-line
+  }, []);
+
+  const changeTheNumberOfServings = useCallback(
+    (changesTo, indexProducts) => () => {
```
Comment (pkarpovich) [inline src/pages/menu/basket/index.js]: diff

File: src/pages/menu/basket/index.js [hunk 24]
```diff
@@ -0,0 +1,381 @@
...
+      if (productsInBasketAtoms[indexProducts].count + changesTo >= 1) {
+        setProductsInBasketAtoms([
```
Comment (pkarpovich) [inline src/pages/menu/basket/index.js]: arrow function

File: src/pages/menu/basket/index.js [hunk 25]
```diff
@@ -0,0 +1,381 @@
...
+          ...productsInBasketAtoms.slice(0, indexProducts),
+          {
+            count: productsInBasketAtoms[indexProducts].count + changesTo,
```
Comment (pkarpovich) [inline src/pages/menu/basket/index.js]: ...item

File: src/pages/menu/basket/index.js [hunk 26]
```diff
@@ -0,0 +1,381 @@
...
+    }
+
+    // eslint-disable-next-line
+  }, []);
+
+  const changeTheNumberOfServings = useCallback(
+    (changesTo, indexProducts) => () => {
+      if (productsInBasketAtoms[indexProducts].count + changesTo >= 1) {
+        setProductsInBasketAtoms([
+          ...productsInBasketAtoms.slice(0, indexProducts),
```
Comment (pkarpovich) [inline src/pages/menu/basket/index.js]: map

File: src/pages/menu/basket/index.js [hunk 27]
```diff
@@ -0,0 +1,381 @@
...
+      ),
+    ]);
+  }, [
+    candidateForDeletion.indexInBasketAtom,
+    productsInBasketAtoms,
+    setProductsInBasketAtoms,
+  ]);
+
+  const sendAnOrder = useCallback(() => {
+    if (!disable && productsInBasketAtoms.length) {
```
Comment (pkarpovich) [inline src/pages/menu/basket/index.js]: isDisabled

File: src/pages/menu/basket/index.js [hunk 28]
```diff
@@ -0,0 +1,381 @@
...
+      fetch(`/api/message/${restaurantId}/${tableId}`, {
```
Comment (pkarpovich) [inline src/pages/menu/basket/index.js:104]: async/await

File: src/pages/menu/expanded-menu/components/navigation/index.js [hunk 29]
```diff
@@ -0,0 +1,48 @@
+import { memo, useCallback } from "react";
+import { useLocation } from "wouter";
+import PropTypes from "prop-types";
+
+import { Flex } from "components/flex";
+import { Text } from "components/text";
+import { theme } from "theme";
+
+const Navigation = (props) => {
```
Comment (pkarpovich) [inline src/pages/menu/expanded-menu/components/navigation/index.js]: {}

**Review by pkarpovich — CHANGES_REQUESTED** [review]

**Review by pkarpovich — APPROVED** [review]

**Pull request merged**